File: src/routes/projects.js
const express = require('express');
const router = express.Router();
const db = require('../models/db');
const { v4: uuidv4 } = require('uuid');

// GET all projects
router.get('/', async (req, res) => {
  try {
    const projects = await db.projects.findAsync({});
    projects.sort((a, b) => new Date(b.createdAt) - new Date(a.createdAt));

    for (const p of projects) {
      p.repoCount = await db.repositories.countAsync({ projectId: p._id });
    }

    res.json({ success: true, data: projects });
  } catch (err) {
    res.status(500).json({ success: false, error: err.message });
  }
});

// GET single project
router.get('/:id', async (req, res) => {
  try {
    const project = await db.projects.findOneAsync({ _id: req.params.id });
    if (!project) return res.status(404).json({ success: false, error: 'Project not found' });
    res.json({ success: true, data: project });
  } catch (err) {
    res.status(500).json({ success: false, error: err.message });
  }
});

// POST create project
router.post('/', async (req, res) => {
  try {
    const { name, description, color, icon } = req.body;
    if (!name || !name.trim()) return res.status(400).json({ success: false, error: 'Name required' });

    const project = await db.projects.insertAsync({
      _id: uuidv4(),
      name: name.trim(),
      description: description?.trim() || '',
      color: color || '#6366f1',
      icon: icon || '📁',
      createdAt: new Date().toISOString(),
      updatedAt: new Date().toISOString(),
    });

    res.json({ success: true, data: project });
  } catch (err) {
    res.status(500).json({ success: false, error: err.message });
  }
});

// PUT update project
router.put('/:id', async (req, res) => {
  try {
    const { name, description, color, icon } = req.body;
    await db.projects.updateAsync(
      { _id: req.params.id },
      { $set: { name, description, color, icon, updatedAt: new Date().toISOString() } }
    );
    const updated = await db.projects.findOneAsync({ _id: req.params.id });
    res.json({ success: true, data: updated });
  } catch (err) {
    res.status(500).json({ success: false, error: err.message });
  }
});

// DELETE project + its repos + docs
router.delete('/:id', async (req, res) => {
  try {
    const repos = await db.repositories.findAsync({ projectId: req.params.id });
    for (const repo of repos) {
      await db.documents.removeAsync({ repositoryId: repo._id }, { multi: true });
    }
    await db.repositories.removeAsync({ projectId: req.params.id }, { multi: true });
    await db.projects.removeAsync({ _id: req.params.id }, {});
    res.json({ success: true });
  } catch (err) {
    res.status(500).json({ success: false, error: err.message });
  }
});

module.exports = router;

File: src/routes/projects.py
"""
src/routes/projects.py
CRUD endpoints for Projects.
"""
import json
import uuid
from datetime import datetime, timezone
from bottle import request, response
from src.models.database import get_conn


def now_iso():
    return datetime.now(timezone.utc).isoformat()


def register(app):
    """Attach all project routes to the Bottle app."""

    @app.get('/api/projects')
    def list_projects():
        with get_conn() as conn:
            rows = conn.execute("""
                SELECT p.id, p.name, p.description, p.icon, p.color,
                       p.created_at, p.updated_at,
                       COUNT(r.id) AS repo_count
                FROM projects p
                LEFT JOIN repositories r ON r.project_id = p.id
                GROUP BY p.id
                ORDER BY p.created_at DESC
            """).fetchall()
        return {'success': True, 'data': [dict(r) for r in rows]}

    @app.get('/api/projects/<project_id>')
    def get_project(project_id):
        with get_conn() as conn:
            row = conn.execute(
                'SELECT * FROM projects WHERE id = ?', (project_id,)
            ).fetchone()
        if not row:
            response.status = 404
            return {'success': False, 'error': 'Project not found'}
        return {'success': True, 'data': dict(row)}

    @app.post('/api/projects')
    def create_project():
        body = request.json or {}
        name = (body.get('name') or '').strip()
        if not name:
            response.status = 400
            return {'success': False, 'error': 'name is required'}
        project_id = str(uuid.uuid4())
        ts = now_iso()
        with get_conn() as conn:
            conn.execute(
                """INSERT INTO projects (id, name, description, icon, color, created_at, updated_at)
                   VALUES (?, ?, ?, ?, ?, ?, ?)""",
                (project_id, name,
                 body.get('description', ''),
                 body.get('icon', '📁'),
                 body.get('color', '#61afef'),
                 ts, ts)
            )
            row = conn.execute('SELECT * FROM projects WHERE id = ?', (project_id,)).fetchone()
        return {'success': True, 'data': dict(row)}

    @app.put('/api/projects/<project_id>')
    def update_project(project_id):
        body = request.json or {}
        name = (body.get('name') or '').strip()
        if not name:
            response.status = 400
            return {'success': False, 'error': 'name is required'}
        ts = now_iso()
        with get_conn() as conn:
            conn.execute(
                """UPDATE projects
                   SET name=?, description=?, icon=?, color=?, updated_at=?
                   WHERE id=?""",
                (name, body.get('description', ''),
                 body.get('icon', '📁'), body.get('color', '#61afef'),
                 ts, project_id)
            )
            row = conn.execute('SELECT * FROM projects WHERE id = ?', (project_id,)).fetchone()
        if not row:
            response.status = 404
            return {'success': False, 'error': 'Project not found'}
        return {'success': True, 'data': dict(row)}

    @app.delete('/api/projects/<project_id>')
    def delete_project(project_id):
        with get_conn() as conn:
            conn.execute('DELETE FROM projects WHERE id = ?', (project_id,))
        return {'success': True}
